Bug fix for mount path handling

ambarve opened the pull request:
Currently when handling 'container_path' elements in container mounts we simply call
filepath.Clean on those paths. However, filepath.Clean adds an extra '.' if the path is a
simple drive letter ('E:' or 'Z:' etc.). Similarly, a drive path like Z:\ is also kept as
it is with an ending '\'. Either of these type of paths cause failures (with incorrect
parameter error) when creating containers via hcsshim. This commit correctly removes the
'.' or '\' at the end of the paths.

Signed-off-by: Amit Barve <ambarve@microsoft.com>

ambarve commented:
@kevpar Please take a look.

ambarve commented:
Fixes https://github.com/containerd/containerd/issues/6589

kevpar commented on pkg/cri/opts/spec_windows.go:
```diff
@@ -142,6 +142,16 @@ func WithWindowsMounts(osi osinterface.OS, config *runtime.ContainerConfig, extr
+				// Remove it here.
+				// No other valid path after filepath.Clean should have only
+				// 3 letters with ':' as the 2nd char.
+				if len(dst) == 3 && dst[1] == ':' {
```
I think it would be clearer to check up above if the path is a drive letter (e.g. `D:`), and skip the `Clean` call if so.

ambarve replied:
I specifically avoided that because someone can still pass a path like `D:\.\.\` which would pass our check of not being a drive letter, go through filepath.Clean and will still cause failures.

kevpar replied:
`D:\.\.\` should clean to `D:\` which I expect would work fine as a mount path. Does that not work?

ambarve replied:
Yeah, even `D:\` would fail with the same error.

kzys commented:
Is `C:\` different from `C:` universally in Windows? I wonder whether we should wrap filepath.Clean or not.

kevpar commented:
> Is `C:\` different from `C:` universally in Windows? I wonder whether we should wrap filepath.Clean or not.

Yes, these are different. `C:\` represents the "root directory" of the C drive, whereas `C:` can have two different meanings based on context:
- It could mean we are talking about the C drive itself. This is the desired meaning here, where we are mapping a directory into the container as a drive that doesn't currently exist.
- To indicate the working directory on the C drive. To understand this, we need to know that for [various weird historical reasons](https://devblogs.microsoft.com/oldnewthing/20101011-00/?p=12563), Windows has (at least in some places) support for actually tracking a working directory _per drive_. In this case the drive letter and colon can be used to refer to a path relative to that drive's working directory. So for instance `D:foo\bar.txt` refers to `foo\bar.txt` relative to the D drive's working directory.
  Presumably for compatibility, Go seems to have carried over this behavior into `filepath.Clean`. In this case it is converting `D:` into `D:.`, which is correct when you consider how this form is used for a relative path.

For our case here, we don't care about relative paths for mounts, and we want to treat the drive letter as a special case, because it allows us to mount a directory as a new drive in the container.

Maybe an argument could be made that we should also support `D:\` to mount to a new drive. I'm not sure if that's a good idea or not.

kzys reviewed (commented):
The change looks good. Can you squash the commits into one?

jterry75 commented:
I'm confused. I'm sure you tested this but I thought a Windows mount had to be to a nonexistent folder? I didn't know you could do:

C:\Path\On\Host -> C: (in the container)

Does it just place the files from `Host` alongside whatever is already present in `C:`? How are conflicts resolved?

ambarve commented:
@jterry75 Good point. Providing `C:` as the destination path fails (with invalid data error) because that's not supported (not sure if there is even a way to properly support it) but we don't have a check for that here. I will add that check.
AFAIK containers don't start with other drives and so this question won't arise for other drives.

kzys commented on pkg/cri/opts/spec_windows.go:
```diff
@@ -138,9 +138,18 @@ func WithWindowsMounts(osi osinterface.OS, config *runtime.ContainerConfig, extr
+					if dst[0] == '\\' {
+						dst = "C:" + dst
+					}
+				} else if dst[0] == 'c' || dst[0] == 'C' {
```
- Wouldn't it be applicable for D: and other drive letters as well?
- We probably should have some tests around this function.

jterry75 replied:
Agreed lets add some tests. To my knowledge there is no destination on Windows for a container that isn't C:\ rooted right? Should we just fail all mounts that don't have this prefix? Is there some type of mount I am not considering? (I know at the os layer there are but at this layer its just a bind mount right?)

jterry75 replied:
Agreed lets add some tests. To my knowledge there is no destination on Windows for a container that isn't C:\ rooted right? Should we just fail all mounts that don't have this prefix? Is there some type of mount I am not considering? (I know at the os layer there are but at this layer its just a bind mount right?)

jterry75 replied:
Why is GH double posting...

ambarve replied:
@kzys It shouldn't fail for drive letters other than `C` since you can mount a directory on the host as a drive inside the container. It is just not possible with the `C:` drive because that drive is already used inside the container when container starts. I have added tests. Let me know if it looks good.
@jterry75 As I said above you can mount a directory on the host as a drive inside the container. Docker supports this but containerd doesn't (due to this bug).

jterry75 replied:
I missed that thanks for the clarity

jterry75 commented on pkg/cri/opts/spec_test.go:
```diff
@@ -44,3 +47,29 @@ func TestOrderedMounts(t *testing.T) {
 	assert.Equal(t, expected, mounts)
 }
+
+func TestDriveMounts(t *testing.T) {
```
I think you need this in spec_windows_test.go

jterry75 commented on pkg/cri/opts/spec_test.go:
```diff
@@ -44,3 +47,29 @@ func TestOrderedMounts(t *testing.T) {
+		expectedContainerPath string
+		expectedError         error
+	}{
+		{&runtime.Mount{HostPath: "C:\\", ContainerPath: "D:\\foo"}, "D:\\foo", nil},
```
I'm good with this but on other PR's I've seen feedback to use literals so `C:\` `D:\foo` rather than `\\`

jterry75 replied:
Do me a solid and add a named pipe mount too to valid that fork of the code.

kevpar replied:
+1 to using raw strings (backticks) so that we don't need to escape backslashes.

jterry75 commented on pkg/cri/opts/spec_windows.go:
```diff
@@ -61,6 +61,61 @@ func cleanMount(p string) string {
+			}
+		}
+		var err error
+		src, err = osi.ResolveSymbolicLink(src)
```
Uh shouldnt this be done before the stat? I know you didnt change this code but still. Does it seem wrong?

jterry75 replied:
Ah ok makes sense.

jterry75 reviewed (approved):
LGTM. The linter looks like its static checks on unchanged files.

kzys commented:
Rebasing against main should fix the linter issue.

ambarve commented:
@kzys & @jterry75 actually I moved the entire `dst` path check out of the named pipe if condition since those checks can & should happen regardless of whether `src` is a pipe or not. I also rebased on master. PTAL and feel free to merge it if it looks good.

jterry75 commented:
Looks like the pipe mount test is what failed. Can you take a look?

kzys reviewed (approved):
LGTM!

jterry75 reviewed (approved):
LGTM

estesp reviewed (approved):
LGTM
